### DISCUSSION TITLE
acquire_token_silent alwayes return None although token_cache is enabled

### DISCUSSION THREAD

**Discussion by karam-jaber-intel** [AzureAD/microsoft-authentication-library-for-python#614] [Q&A]
I'm trying to use persistent cache using the local storage by initializing token_cache but acquire_token_silent  alwayes return None although the binary for the cache exist locally.

Here is my code:

```
import os, atexit, msal
cache = msal.SerializableTokenCache()
if os.path.exists("my_cache.bin"):
    cache.deserialize(open("my_cache.bin", "r").read())
atexit.register(lambda:
    open("my_cache.bin", "w").write(cache.serialize())
    # Hint: The following optional line persists only when state changed
    if cache.has_state_changed else None
    )
# Create an instance of the PublicClientApplication
app = PublicClientApplication(client_id, authority=authority,token_cache=cache)

# Attempt to acquire a token silently with verbose logging
result = app.acquire_token_silent(["User.Read"], account=None)

if result is not None and "access_token" in result:
  access_token = result["access_token"]
else:
  user_email = input('User email: ') 
  result = app.acquire_token_by_username_password(user_email,
                                                                    getpass('Password: '),
                                                                    scopes=["User.Read"])
  access_token = result["access_token"]
response = requests.get(request_url, headers=headers, verify=False)
```

any idea what Im missing ?

**Comment by karam-jaber-intel** [comment]
found the issue that starting from msal 1.23 account=None input to acquire_token_silent is considered No-OP and alwayes returns NULL

**Reply by rayluo** [reply]
Incorrect. You are mixing things up. Acquiring token for a user (such as this [username password sample](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/dev/sample/username_password_sample.py)) and acquiring token for an app (such as this [confidential client sample](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/dev/sample/confidential_client_secret_sample.py)) have two different usage pattern.

**Reply by karam-jaber-intel** [reply]
but the code worked once i changed this 
`result = app.acquire_token_silent(["User.Read"], account=None)`
to this ;
`result = app.acquire_token_silent(["User.Read"], account=self.app.get_accounts()[0])`

are you claiming that the above code is logically not correct ?

**Reply by rayluo** [reply]
The calling patterns demonstrated in our official samples have always been two different pairs.

1. When acquiring token for an end user, [app developer shall first call `acquire_token_silent(..., account=some_account)` which will utilize token cache; should it be a cache miss, you then call one of the acquire token *for user* methods such as `acquire_token_interactive()`](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/dev/sample/interactive_sample.py#L61-L66).
2. When acquiring token for your current app (not for a user), app developer used to call `acquire_token_silent(..., account=None)` and then an `acquire_token_for_client()`. (Since MSAL 1.23, this calling pattern has been simplified to only `acquire_token_for_client()`.)

@karam-jaber-intel , your initial snippet was mix-and-matching the methods, and your calling pattern did not match either of the two pairs above.

> are you claiming that the above code is logically not correct ?

Your latest change was correct, but your understanding was incorrect. It is not because of MSAL 1.23 that you would need to make such a change, it was just that your original calling pattern wasn't right in the first place. MSAL 1.23+ prevents your incorrect calling pattern from doing more harm.
